xds/internal/balancer/cdsbalancer/cdsbalancer_test.go

```diff
@@ -315,507 +322,1272 @@ func setupWithWatch(t *testing.T) (*fakeclient.Client, *cdsBalancer, *testEDSBal
-	defer sCancel()
-	if err := edsB.waitForClientConnUpdate(sCtx, balancer.ClientConnState{}); err != context.DeadlineExceeded {
-		t.Fatalf("ClusterUpdate after close forwaded to EDS balancer")
+	// Create a manual resolver and push a service config specifying the use of
+	// the cds LB policy as the top-level LB policy, and a corresponding config
+	// with a single cluster.
+	r := manual.NewBuilderWithScheme("whatever")
+	jsonSC := fmt.Sprintf(`{
+			"loadBalancingConfig":[{
+				"cds_experimental":{
+					"cluster": "%s"
+				}
+			}]
+		}`, clusterName)
+	scpr := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(jsonSC)
+	r.InitialState(xdsclient.SetClient(resolver.State{ServiceConfig: scpr}, xdsC))
+
+	// Create a ClientConn, and this should trigger the cds LB policy.
+	cc, err := grpc.Dial(r.Scheme()+":///test.service", grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(r))
+	if err != nil {
+		t.Fatalf("Failed to dial local test server: %v", err)
 	}
+	t.Cleanup(func() { cc.Close() })
 
-	// Make sure that the underlying EDS balancer is closed.
-	if err := edsB.waitForClose(ctx); err != nil {
-		t.Fatal(err)
+	// Verify that a successful RPC can be made.
+	client := testgrpc.NewTestServiceClient(cc)
+	if _, err = client.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true)); err != nil {
+		t.Fatalf("EmptyCall() failed: %v", err)
 	}
 
-	// Make sure that the UpdateClientConnState() method on the CDS balancer
-	// returns error.
-	if err := cdsB.UpdateClientConnState(cdsCCS(clusterName, xdsC)); err != errBalancerClosed {
-		t.Fatalf("UpdateClientConnState() after close returned %v, want %v", err, errBalancerClosed)
+	// Retrieve the cds LB policy policy and close it.
+	var cdsBal balancer.Balancer
+	select {
+	case cdsBal = <-cdsBalancerCh:
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for cds LB policy to be created")
 	}
+	cdsBal.Close()
 
-	// Make sure that the ResolverErr() method on the CDS balancer does not
-	// forward the update to the EDS balancer.
-	rErr := errors.New("cdsBalancer resolver error")
-	cdsB.ResolverError(rErr)
-	sCtx, sCancel = context.WithTimeout(context.Background(), defaultTestShortTimeout)
-	defer sCancel()
-	if err := edsB.waitForResolverError(sCtx, rErr); err != context.DeadlineExceeded {
-		t.Fatal("ResolverError() forwarded to EDS balancer after Close()")
+	// Wait for the CDS resource to be not requested anymore.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for CDS resource to not requested")
+	case <-cdsResourceCanceledCh:
+	}
+	// Wait for the child policy to be closed.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for the child policy to be closed")
+	case <-childPolicyCloseCh:
 	}
 }
 
+// Tests the calling ExitIdle on the cds LB policy results in the call being
+// propagated to the child policy.
 func (s) TestExitIdle(t *testing.T) {
```

Comment:
Honestly, this test seems like a huge test to test that the ExitIdle() call propagates. Up to you, but I wouldn't mind deleting this test.

Reply:
It's much shorter now. And I think it is an important test to have.